From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: qIooIp <dino.kanizaj@gmail.com>
Date: Sat, 5 Oct 2024 22:09:22 +0200
Subject: [PATCH] Option to disable firework damage


diff --git a/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java b/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
index 7005b3a26e9b5d79064981a4a41bee21b65a9fc3..dc0063beb8101c9c2131ce711e018741a4cfe540 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
@@ -4,6 +4,8 @@ import java.util.Iterator;
 import java.util.List;
 import java.util.OptionalInt;
 import javax.annotation.Nullable;
+
+import com.ngxdev.SilverConfig;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.particles.ParticleTypes;
 import net.minecraft.nbt.CompoundTag;
@@ -187,7 +189,13 @@ public class FireworkRocketEntity extends Projectile implements ItemSupplier {
     private void explode() {
         this.level().broadcastEntityEvent(this, (byte) 17);
         this.gameEvent(GameEvent.EXPLODE, this.getOwner());
-        this.dealExplosionDamage();
+
+        // Floki start
+        if(SilverConfig.fireworkDamage) {
+            this.dealExplosionDamage();
+        }
+        // Floki end
+
         this.discard();
     }
 
